### PR TITLE
Replace deprecated `requests` with `node-fetch`

### PR DESCRIPTION
**Pull request by rmeritz**
- Requests has been deprecated: https://github.com/request/request/issues/3142

- There are lot of options for replacements: https://github.com/request/request/issues/3143

- Went with `node-fetch` because it is the most lightweight: https://github.com/node-fetch/node-fetch

- Use abort-controller to add back the timeout
https://github.com/node-fetch/node-fetch#request-cancellation-with-abortsignal

- When we get a request back cancel the timeout because we might need to
recur. https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/clearTimeout

- Tested that `node-fetch` replacement actually works end-to-end by running the `test/test_files/index.html`

- In order to test against the latest changes I regenerated the`browserify`ed js with `node_modules/grunt/bin/grunt browserify` and copied the built file to the test directly. I'm unclear what the process should be for when `airtable.browser.js` should be updated. I didn't check in the built version, because I assume that is only updated when I release is done. And explanation of how and when to regenerate the `airtable.browser.js` would be great to add to the README, but I'm not sure what the process is.

### CODE REVIEW

**Comment by rmeritz** [comment]
So it seems that `node-fetch` doesn't work with node 8 which is now in LTS. https://nodejs.org/en/blog/release/v8.9.0/

I've replicated all the falling tests on the node 8 build locally and it is a real problem with the library.

Looking at the commit history it seems like this library had only be supporting node 10+ but then a decision was explicitly made to support node 8+. https://github.com/Airtable/airtable.js/commit/31fd0a089fee87832760f35c7270eae283972e35

@kasrak - Do we definitely want to support node 8 moving forward? I can close this in favor of another solution if so. Sorry I didn't catch this issue earlier, but at least we have CI!

**Comment by rmeritz** [comment]
@kasrak - So it turns out the probably wasn't with `node-fetch` but with my own code changes because I was using `finally` which wasn't in node 8. So you can ignore what I said about node 8. This should be good to go now.

**Comment by kasrak** [comment]
> In order to test against the latest changes I regenerated thebrowserifyed js with node_modules/grunt/bin/grunt browserify and copied the built file to the test directly. I'm unclear what the process should be for when airtable.browser.js should be updated. I didn't check in the built version, because I assume that is only updated when I release is done. And explanation of how and when to regenerate the airtable.browser.js would be great to add to the README, but I'm not sure what the process is.

Yep, the workflow is to manually generate a new airtable.browser.js as part of a release (e.g. https://github.com/Airtable/airtable.js/commit/31fd0a089fee87832760f35c7270eae283972e35). But in this case let's keep checking in the built bundle to verify the size as part of this code review. Once we're happy with it, we'll uncheck-in those changes and land this PR without any changes to airtable.browser.js

**Comment by rmeritz** [comment]
@kasrak - I'm working on figuring out a solution to this. In order to move forward it would be helpful to know what browsers (and versions) this library should support. It would help me figure out what polyfills I need to add (or not) and it would help future users have a clear understanding of if the library will work for them or not.

I see we are already polyfilling promise even though it has wide support (except for IE). https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#Browser_compatibility Should that be taken as an explicate decision to support at least IE?

I need to determine if I need to polyfill `fetch` and `abortController` as well. https://developer.mozilla.org/en-US/docs/Web/API/AbortController https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API#Browser_compatibility

**Comment by rmeritz** [comment]
@kasrak - I figured out what was causing the bloat of the built package size. It turns out there was a second use of the `request` package in `base#makeRequest` that I was missing. And even though I had removed the `request` library from package.json browserify was pulling it back in. I've removed the use of `request` in `base#makeRequest` with `fetch` as well. And now the built library is back down to 220K. 

This is now working but it would still be great to address the question I asked about what browser versions we want to support. I didn't end up pulling in pollyfills for `fetch` or `abortController` because they are both widely supported and I didn't want to add bloat. It would be very helpful to me and all future users and maintainers to clearly lay out in the README what versions are supported. 

Can you clarify what versions we support so I can determine if we need to add polyfills or not?

Maybe that would also allow us to pull out the Promise polyfill library as one user already asked for. https://github.com/Airtable/airtable.js/issues/179 I would be happy to make that change as well. 

Also, one reason I missed the use of `request` in `base#makeRequest` is that `makeRequest` is never actually used anywhere. It is technically possible to access it but 100% of API calls are filtered through `runAction` which does practically the same thing as `makeRequest`. Can I just remove `makeRequest` or deprecate it? Do we consider it a public API? I cannot imagine a user wanting to use it and it isn't part of the documentation.

Speaking of the public API, when working making changes to this library and writing tests it came up over and over again that there isn't a clear distinction documented anywhere of what is a public and what is a private interface. I think it would very helpful to the long term maintainability of the project to provide clarity around this. I would be more than happy to add jdocs to make this clear (using the public API docs as my guide) if you agree that that would be a useful path forward. Do you think it makes sense for me to go ahead and do this? (Esp ahead of the Typescript changes).

**Comment by rmeritz** [comment]
Also, I cannot add linked issue, but this should close https://github.com/Airtable/airtable.js/issues/153 when merged.

**Comment by kasrak** [comment]
Thanks!

> This is now working but it would still be great to address the question I asked about what browser versions we want to support. I didn't end up pulling in pollyfills for fetch or abortController because they are both widely supported and I didn't want to add bloat. It would be very helpful to me and all future users and maintainers to clearly lay out in the README what versions are supported.

Let's match the supported browser list for Airtable (https://support.airtable.com/hc/en-us/articles/217990018), which is currently:
- Chrome 67 or higher
- Firefox 52 or higher
- Safari 10 or higher
- Edge 25 or higher

So we can drop support for IE, drop the Promise polyfill (maybe in a separate PR?), and do a major version bump when we release the next version of Airtable.js

> Can I just remove makeRequest or deprecate it? Do we consider it a public API?

Let's leave it in for now, it's handy for making more custom requests to the API.

> there isn't a clear distinction documented anywhere of what is a public and what is a private interface.

By convention, any identifiers beginning with `_` are private. Let's chat separately about ways to make that more clear.

**Comment by rmeritz** [comment]
> Does any of the above require changes to this PR? Let me know if not and I'll review the code changes. Thanks!

Yes. I have to make one more set of changes. Thanks so much for providing clear requirements on browser support. It does mean that I will need to add 2 polyfills though: one for `fetch` just for Safari 10.0 and one for `AbortController` for Safari 10 - 11 and Firefox 52 - 56. 

> drop the Promise polyfill (maybe in a separate PR?), and do a major version bump when we release the next version of Airtable.js

I'm drop the promise polyfill in a follow-up PR.

> By convention, any identifiers beginning with _ are private. Let's chat separately about ways to make that more clear.

I think that is quite clear. But the API documented in the developer docs for airtable.js (the ones generated per base) only shows the use of a tiny tiny percentage of what according to the convention is technically public. It would be helpful to the long-term maintainability of the library to only commit to a small externally public API -- meaning that there would be a distinction between methods that a are private to a file (using underscore) and methods that are private to external use. Anyway, this definitely doesn't have to be answered in this PR at all. I just thought I would mention it while it was top of mind.

**Comment by rmeritz** [comment]
@kasrak - I added the polyfill needed and tested it works with Browserstack in the older browsers, so you can review it again.

File: test/test_files/airtable.browser.js [hunk 1]
```diff
@@ -1,4 +1,17 @@
 require=(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
```
Comment (minicat) [inline test/test_files/airtable.browser.js:1]: remember to revert these changes before landing!

File: test/base.test.js [hunk 2]
```diff
@@ -346,32 +355,26 @@ describe('Base', function() {
             });
 
             it('retries 429s until success', function() {
-                const realSetTimeout = setTimeout;
-
-                jest.useFakeTimers();
```
Comment (minicat) [inline test/base.test.js:351]: why was this changed? & number of requests changed from 3 to 2

File: lib/base.js [hunk 3]
```diff
@@ -34,59 +34,76 @@ Base.prototype.makeRequest = function(options) {
...
-        qs: get(options, 'qs', {}),
         headers: this._getRequestHeaders(get(options, 'headers', {})),
-        json: true,
-        timeout: this._airtable.requestTimeout,
+        signal: controller.signal,
     };
+
     if ('body' in options && _canRequestMethodIncludeBody(method)) {
-        requestOptions.body = options.body;
+        requestOptions.body = JSON.stringify(options.body);
```
Comment (minicat) [inline lib/base.js:56]: im assuming the `JSON.stringify` was previously handled by the request library?

File: test/test_helpers.js [hunk 4]
```diff
@@ -143,6 +143,7 @@ function getMockEnvironmentAsync(options) {
                         airtable: new Airtable({
                             apiKey: 'key123',
                             endpointUrl: 'http://localhost:' + testServerPort,
+                            requestTimeout: 1000,
```
Comment (minicat) [inline test/test_helpers.js:146]: nit: maybe add a comment mentioning this is required for the timeout tests

File: lib/base.js (hunk 3 above)
Comment (rmeritz) [reply lib/base.js:56]: It was.

File: README.md [hunk 5]
```diff
@@ -18,6 +18,7 @@ To install airtable.js in a node project:
 
     npm install airtable
 
+Airtable.js should work with Node 10 and above.
```
[the remark this answers is not in the document]
Comment (rmeritz) [reply README.md:21]: Good call.

File: README.md [hunk 6]
```diff
@@ -33,6 +34,8 @@ Edit `test/test_files/index.html` - put your `BASE_ID` and `API_KEY` (Be careful
 
 Then open http://localhost:8000/ in your browser.
 
+Airtable.js should work with browsers supported by Airtable App.
```
[the remark this answers is not in the document]
Comment (rmeritz) [reply README.md:37]: Fixed as well.

File: test/test_files/airtable.browser.js (hunk 1 above)
Comment (rmeritz) [reply test/test_files/airtable.browser.js:1]: I've reverted it.

File: test/test_helpers.js (hunk 4 above)
Comment (rmeritz) [reply test/test_helpers.js:146]: Done.

File: test/base.test.js (hunk 2 above)
Comment (rmeritz) [reply test/base.test.js:351]: It was a few weeks ago, so I'm not 100% sure. But I think I just changed it from 2 to 3 requests to make the tests faster.

As work that procceded, I wrote a bunch of new timeout tests (to get 100% coverage) that use the actual timeouts and I think I just made this test more consistent with the other. 

I think I did this work because I ran into an issue with how I passed the timeout signal with the new node fetch library which caused me to dig in with these tests more.

**Comment by rmeritz** [comment]
@minicat - Thanks for taking a look. I think I addressed all your comments and answered your questions now.

File: README.md [hunk 7]
```diff
@@ -18,7 +18,7 @@ To install airtable.js in a node project:
 
     npm install airtable
 
-Airtable.js should work with Node 10 and above.
+Airtable.js is compatible work with Node 10 and above.
```
Comment (minicat) [inline README.md:21]:
```suggestion
Airtable.js is compatible with Node 10 and above.
```

File: README.md [hunk 8]
```diff
@@ -34,7 +34,7 @@ Edit `test/test_files/index.html` - put your `BASE_ID` and `API_KEY` (Be careful
 
 Then open http://localhost:8000/ in your browser.
 
-Airtable.js should work with browsers supported by Airtable App.
+Airtable.js is compatiable with browsers supported by Airtable App.
```
Comment (minicat) [inline README.md:37]:
```suggestion
Airtable.js is compatible with browsers supported by the Airtable web app.
```

File: test/base.test.js (hunk 2 above)
Comment (minicat) [reply test/base.test.js:351]: Ahh i see - sounds reasonable, thanks!

**Comment by rmeritz** [comment]
@minicat - I committed spelling/grammar changes you caught. Thanks!
